descriptors: improve descriptor cache and cache xpubs

achow101 opened the pull request:
Improves the descriptor cache by changing it from a `std::vector<unsigned char>` to a newly introduced `DescriptorCache` class. Instead of serializing pubkeys and whatever else we would want to cache in a way that may not be backwards compatible, we instead create a `DescriptorCache` object and populate it. This object contains only an xpub cache. Since the only `PubkeyProvider` that used the cache is the `BIP32PubkeyProvider` we just have it store the xpubs instead of the pubkeys. This allows us to have both the parent xpub and the child xpubs in the same container. The map is keyed by `KeyOriginInfo`.

Sine we are caching `CExtPubKey`s in `DescriptorCache`, `BIP32PubKeyProviders` can use the cached parent xpubs to derive the children if unhardened derivation is used in the last step. This also means that we can still derive the keys for a `BIP32PubkeyProvider` that has hardened derivation steps. When combined with descriptor wallets, this should allow us to be able to import a descriptor with an `xprv` and hardened steps and still be able to derive from it. In that sense, this is an alternative to #18163

To test that this works, the tests have been updated to do an additional `Expand` at the `i + 1` position. This expansion is not cached. We then do an `ExpandFromCache` at `i + 1` and use the cache that was produced by the expansion at `i`. This way, we won't have the child xpubs for `i + 1` but we will have the parent xpubs. So this checks whether the parent xpubs are being stored and can be used to derive the child keys. Descriptors that have a hardened last step are skipped for this part of the test because that will always require private keys.

achow101 committed:
Introduce DescriptorCache struct which caches xpubs

Sjors commented on src/script/descriptor.h:
```diff
@@ -13,6 +13,18 @@
+    DescriptorCache() : m_pos_begin(0) {}
+    DescriptorCache(int begin_pos) : m_pos_begin(begin_pos) {}
+    bool CachePubKey(unsigned int cache_pos, const CPubKey& pubkey);
+    bool GetCachedPubKey(unsigned int global_pos, unsigned int internal_pos, CPubKey& pubkey) const;
```
`global_pos` and `internal_pos` could use documentation. Would it make sense to cache all pubkeys (for a given position) in one operation?

Sjors reviewed (commented):
Concept ACK.

I'm seeing errors on Travis and locally:

```
AssertionError: Unexpected stderr Assertion failed: (fValid), function GetPubKey, file key.cpp, line 185.
```

instagibbs reviewed (commented):
this is getting importmulti test failures

achow101 commented:
Fixed the assertion. Travis seems to be seeing a memory leak but I'm having trouble figuring out where that is.

achow101 committed:
Add DescriptorCache* read_cache and DescriptorCache* write_cache to Expand and GetPubKey

Have Expand, ExpandFromCache, and ExpandHelper take additional DescriptorCache
parameters. These are then passed into PubkeyProvider::GetPubKey which
also takes them as arguments.

Reading and writing to the cache is pushed down into GetPubKey. The old cache where
pubkeys are serialized to a vector is completely removed and instead xpubs are being
cached in DescriptorCache.

achow101 committed:
Cache the immediate derivation parent xpub

If unhardened derivation is used, cache the immediate derivation
parent xpub and use it for unhardened derivation

achow101 commented:
I've reworked how the caching works to get rid of the `global_pos` and `internal_pos` stuff.

Instead of having `ExpandHelper` go through the cache and find the pubkeys, we pass the cache into the `PubkeyProvider`s and let them do the cache lookups. Since really only `BIP32PubkeyProvider` uses the cache, we can get rid of the position stuff by having it store only xpubs in a map. So for all derived keys, we store their `KeyOriginInfo` and xpub so that later the `BIP32PubkeyProvider` can look them up. The last parent is still being stored so that additional derivation can occur.

This consolidates cache reads and writes and encapsulates them in the classes that actually use the cache.

Sjors commented on src/test/descriptor_tests.cpp:
```diff
@@ -135,19 +136,34 @@ void DoCheck(const std::string& prv, const std::string& pub, int flags, const st
+            if (!(flags & DERIVE_HARDENED)) {
+                // Evaluate the descriptor at i + 1
+                FlatSigningProvider script_provider1, script_provider_cached1;
+                std::vector<CScript> spks1, spks_cached1;
```
nit: maybe rename `spks_cached1` to `spks1_from_cache`

achow101 replied:
Done

Sjors commented on src/test/descriptor_tests.cpp:
```diff
@@ -135,19 +136,34 @@ void DoCheck(const std::string& prv, const std::string& pub, int flags, const st
+                std::vector<CScript> spks1, spks_cached1;
+                BOOST_CHECK((t ? parse_priv : parse_pub)->Expand(i + 1, key_provider, spks1, script_provider1, nullptr));
+
+                // Try again but use the old cache. The old cache won't have the pubkeys, but will have the parent xpub for derivation.
```
`but use the cache generated while expanding i. This cache won't have the pubkeys for i + 1, but`

achow101 replied:
Done

Sjors commented on src/script/descriptor.cpp:
```diff
@@ -262,6 +262,16 @@ class BIP32PubkeyProvider final : public PubkeyProvider
         return true;
     }
 
+    // Derives the last xpub
```
`xpub or xpriv`, unless you neuter it at the end and have it return a `CExtPubKey&`

achow101 replied:
Changed to xprv

Sjors commented on src/script/descriptor.cpp:
```diff
@@ -275,29 +285,49 @@ class BIP32PubkeyProvider final : public PubkeyProvider
+            if (write_cache && !read_cache) {
+                if (!write_cache->CacheExtPubKey(info_out, extkey)) return false;
+                // Only cache parent if there is any unhardened derivation
+                if (m_derive == DeriveType::UNHARDENED && !write_cache->CacheExtPubKey(parent_info, parent)) return false;
```
Is it worth skipping the parent `write` once it's done?

achow101 replied:
A lookup is just as expensive as an insertion, so I don't think so.

Sjors reviewed (commented):
a544f0e7459f346a1bf129c9dc978fe46c6abf73 looks correct. Much clearer than before, but still some confusing bits.

I find the separation of 4f24907f4007b2786ee39dcb8d6f86e52494db10 a6feb5d1aee6ae08b87029561873471cdd62a36a confusing, with the `ExpandFromCache` signature changing in two steps. 

Switching from pubkeys to xpubs plus origin info increases the cache size by ~40 bytes per key, or 240 KB for a typical descriptor wallet with change & receive for three output types, with the first 1000 keys expanded. That's probably acceptable, though maybe the serialization could be optimized to leave out the chain code for "leaf" keys (meh, complexity).

This could be a good time to rename `m_derive` to `m_range_derive` to make it clear that this refers to whether a given (final?) path element is (hardened) ranged or not.

achow101 commented on src/script/keyorigin.h:
```diff
@@ -18,6 +18,21 @@ struct KeyOriginInfo
+            if (a.fingerprint[i] < b.fingerprint[i]) return true;
+            if (a.fingerprint[i] > b.fingerprint[i]) return false;
+        }
+        if (a.path.size() < b.path.size()) return true;
```
Done

achow101 commented:
I've squashed the middle three commits together (the ones that just changed to using DescriptorCache) so that the signature of `Expand` and `ExpandHelper` don't change so many times.

> Switching from pubkeys to xpubs plus origin info increases the cache size by ~40 bytes per key, or 240 KB for a typical descriptor wallet with change & receive for three output types, with the first 1000 keys expanded. That's probably acceptable, though maybe the serialization could be optimized to leave out the chain code for "leaf" keys (meh, complexity).

We could do what I did originally and have a pubkey cache, but I felt that to be unnecessary. We would still be storing `KeyOriginInfo`.

> This could be a good time to rename `m_derive` to `m_range_derive` to make it clear that this refers to whether a given (final?) path element is (hardened) ranged or not.

Why? I find it to be clear.

achow101 committed:
Only cache xpubs that have a hardened last step

Also adds tests for this:
For ranged descriptors with unhardened derivation, we expect to
find parent keys in the cache but no child keys.

For descriptors containing an xpub but do not have unhardened derivation
(i.e. hardened derivation or single xpub with or without derivation),
we expect to find all of the keys in the cache, and the same
number of keys in the cache as in the SigningProvider.

For everything else (no xpub), nothing should be cached at all.

promag commented on src/script/keyorigin.h:
```diff
@@ -18,6 +18,22 @@ struct KeyOriginInfo
+            if (a.path.at(i) > b.path.at(i)) return false;
+        }
+        if (a.path.size() < b.path.size()) return true;
+        if (a.path.size() > b.path.size()) return false;
```
Can be removed.

achow101 replied:
Done

promag commented on src/script/descriptor.h:
```diff
@@ -13,6 +13,15 @@
+    std::map<KeyOriginInfo, CExtPubKey> m_xpubs;
+
+public:
+    bool CacheExtPubKey(const KeyOriginInfo& origin_info, const CExtPubKey& xpub);
```
Will this return false in a follow up change?

achow101 replied:
No. I've made it `void`

promag commented on src/script/descriptor.cpp:
```diff
@@ -1050,3 +1061,16 @@ std::unique_ptr<Descriptor> InferDescriptor(const CScript& script, const Signing
+
+void DescriptorCache::CacheExtPubKey(const KeyOriginInfo& origin_info, const CExtPubKey& xpub)
+{
+    m_xpubs[origin_info] = xpub;
```
Does it make sense to `assert(m_xpubs.count(origin_info) == 0)`?

Sjors replied:
Not at the moment: https://github.com/bitcoin/bitcoin/pull/18204#discussion_r384655169

promag replied:
Performance penalty aside, makes sense to assert it's the same xpub if already cached?

Sjors commented on src/script/descriptor.cpp:
```diff
@@ -275,29 +285,49 @@ class BIP32PubkeyProvider final : public PubkeyProvider
-                if (!GetPrivKey(pos, arg, priv_key)) return false;
-                *key = priv_key.GetPubKey();
+            CExtPubKey extkey = m_extkey;
+            CExtPubKey parent = extkey;
```
This `CExtPubKey` juggling still confuses me :-)

I added some comments for my own sanity. Will probably need to read this again later.

```
CExtPubKey extkey_out = m_extkey;    // to hold extended key for *key (m/.../k)
CExtPubKey extkey_parent = m_extkey; // to hold extended key for the parent of *key (m/.../k/pos)
```

Sjors commented on src/script/descriptor.cpp:
```diff
@@ -275,29 +285,49 @@ class BIP32PubkeyProvider final : public PubkeyProvider
-                *key = priv_key.GetPubKey();
+            CExtPubKey extkey = m_extkey;
+            CExtPubKey parent = extkey;
+            if (read_cache) {
```
```
// Check if m/.../k/pos` is cached (only the case for unranged descriptors)
```

Sjors commented on src/script/descriptor.cpp:
```diff
@@ -275,29 +285,49 @@ class BIP32PubkeyProvider final : public PubkeyProvider
+            if (read_cache) {
+                if (!read_cache->GetCachedExtPubKey(info_out, extkey)) {
+                    if (m_derive == DeriveType::HARDENED) return false;
+                    // Try to get the derivation parent
```
```
// Check if the parent at m/.../k is cached
```

Sjors commented on src/script/descriptor.cpp:
```diff
@@ -275,29 +285,49 @@ class BIP32PubkeyProvider final : public PubkeyProvider
+                    parent.Derive(extkey, pos);
+                }
+            } else if (IsHardened()) {
+                CExtKey xprv;
```
```
// Get the extended private key at m/.../k
```

Sjors commented on src/script/descriptor.cpp:
```diff
@@ -275,29 +285,49 @@ class BIP32PubkeyProvider final : public PubkeyProvider
+            } else if (IsHardened()) {
+                CExtKey xprv;
+                if (!GetDerivedExtKey(arg, xprv)) return false;
+                parent = xprv.Neuter();
```
```
// Get the extended public key at m/.../k/i or m/.../k/i'
```

Sjors commented on src/script/descriptor.cpp:
```diff
@@ -275,29 +285,49 @@ class BIP32PubkeyProvider final : public PubkeyProvider
-                CExtPubKey extkey = m_extkey;
                 for (auto entry : m_path) {
-                    extkey.Derive(extkey, entry);
+                    parent.Derive(parent, entry);
```
(one line up)
```
// Set parent to the extended public key at m/.../k
```

Sjors commented on src/script/descriptor.cpp:
```diff
@@ -275,29 +285,49 @@ class BIP32PubkeyProvider final : public PubkeyProvider
+                    parent.Derive(parent, entry);
                 }
-                if (m_derive == DeriveType::UNHARDENED) extkey.Derive(extkey, pos);
+                extkey = parent;
```
(above)
```
// If this is not a ranged descriptor, we return the public key at `m/.../k`, otherwise at `m/.../k/i
```

Sjors commented:
> We could do what I did originally and have a pubkey cache, but I felt that to be unnecessary.

That's fine. Unless future benchmarks show that `DescriptorScriptPubKeyMan::TopUp()` is particularly slow (https://github.com/bitcoin/bitcoin/pull/16528/commits/76a355eff5b55df6951d21e66f4558ee26f244f4), compared to the other work that happens when loading a wallet.

Can you move 568a90f908313466f5a57c46dc3a16cc01117f70 and 23aef1932c7f8f580928eac595ad3928f306ae35 from #15528 to here? 568a90f908313466f5a57c46dc3a16cc01117f70 could use a test to show that nothing is added to cache.

Sjors reviewed (commented):
> We could do what I did originally and have a pubkey cache, but I felt that to be unnecessary.

That's fine. Unless future benchmarks show that `DescriptorScriptPubKeyMan::TopUp()` is particularly slow (https://github.com/bitcoin/bitcoin/pull/16528/commits/76a355eff5b55df6951d21e66f4558ee26f244f4), compared to the other work that happens when loading a wallet.

Can you move 568a90f908313466f5a57c46dc3a16cc01117f70 and 23aef1932c7f8f580928eac595ad3928f306ae35 from #15528 to here? 568a90f908313466f5a57c46dc3a16cc01117f70 could use a test to show that nothing is added to cache.

instagibbs commented:
Unable to "review" at the moment due to github issues, but https://github.com/bitcoin/bitcoin/pull/18204/commits/d90075e586b2a4b1fbc1c511d662c0b918023595 looks logically correct (will swing back here when back up and drop comments)

achow101 commented on src/script/descriptor.cpp:
```diff
@@ -287,32 +287,44 @@ class BIP32PubkeyProvider final : public PubkeyProvider
             *key = extkey.pubkey;
             if (write_cache && !read_cache) {
                 write_cache->CacheExtPubKey(info_out, extkey);
+                // Only cache parent if there is any unhardened derivation
```
Just the last range step. The range derivation type has to be unhardened or none. This means it either ends with `*` or with an absolute path. There can be hardened steps in between, and even have the last step be hardened if it is not a range.

Sjors replied:
Just to clarify, there's no test for this AFAIK, but `xpriv/../*/...` isn't allowed, right?

instagibbs replied:
I see, so we don't do any intermediate caching if we can always generate from "root" anyways with pubkey.

achow101 replied:
> Just to clarify, there's no test for this AFAIK, but `xpriv/../*/...` isn't allowed, right?

Yes. Maybe we should add a `CheckUnparsable` for that.

> I see, so we don't do any intermediate caching if we can always generate from "root" anyways with pubkey.

Actually no. We always cache if there are xpubs. If the descriptor has just an xpub, e.g. `wpkh(xpub...)`, we actually would cache that xpub. If it were `wpkh(xpub.../0)`, we would cache the derived xpub. Same if it is hardened.

instagibbs replied:
I think I'm agreeing with you here.

achow101 commented on src/script/descriptor.cpp:
```diff
@@ -153,7 +153,7 @@ struct PubkeyProvider
 
     /** Derive a public key. If key==nullptr, only info is desired. */
-    virtual bool GetPubKey(int pos, const SigningProvider& arg, CPubKey* key, KeyOriginInfo& info) const = 0;
+    virtual bool GetPubKey(int pos, const SigningProvider& arg, CPubKey* key, KeyOriginInfo& info, const DescriptorCache* read_cache = nullptr, DescriptorCache* write_cache = nullptr) const = 0;
```
Added some comments.

achow101 commented:
> Can you move 568a90f and 23aef19 from #16528 to here? 568a90f could use a test to show that nothing is added to cache.

Done. Also squashed a bit and added tests.

Sjors commented:
Travis didn't run. Could be because of the Github problems yesterday, or could be if your account was flagged.

Sjors commented on src/script/descriptor.cpp:
```diff
@@ -152,8 +152,11 @@ struct PubkeyProvider
 
-    /** Derive a public key. If key==nullptr, only info is desired. */
-    virtual bool GetPubKey(int pos, const SigningProvider& arg, CPubKey* key, KeyOriginInfo& info) const = 0;
+    /** Derive a public key. If key==nullptr, only info is desired.
```
`CPubKey* key` can be a reference now, and you can drop `if(key)` in a few places. The `nullptr` was only used for caching individual pubkeys.

achow101 replied:
Done

Sjors reviewed (commented):
Keeping you busy :-)

I like the progression in the commits now: b434a8769956573b2ee8eaa2b4826f0e82d39f1f maintains the existing behaviour of caching individual pub keys, except that it switches to xpubs. The next commits then reduce the amount of stuff cached.

achow101 commented on src/script/descriptor.cpp:
```diff
@@ -275,29 +288,37 @@ class BIP32PubkeyProvider final : public PubkeyProvider
-    bool GetPubKey(int pos, const SigningProvider& arg, CPubKey* key, KeyOriginInfo& info) const override
+    bool GetPubKey(int pos, const SigningProvider& arg, CPubKey* key, KeyOriginInfo& info, const DescriptorCache* read_cache = nullptr, DescriptorCache* write_cache = nullptr) const override
     {
+        KeyOriginInfo info_out;
```
We want to avoid modifying `info` in case later steps fail.

instagibbs replied:
might be worth a comment(took me a bit to convince myself what it was doing)

Sjors replied:
Then we need another test, because everything passed when I changed that.

achow101 replied:
I added a comment. I don't think a test is useful. Everything still passes because `ExpandHelper` does the same thing. Just keeping it for belt-and-suspenders.

achow101 commented on src/script/descriptor.cpp:
```diff
@@ -262,6 +265,16 @@ class BIP32PubkeyProvider final : public PubkeyProvider
+    {
+        if (!GetExtKey(arg, xprv)) return false;
+        for (auto entry : m_path) {
+            xprv.Derive(xprv, entry);
```
Done here and elsewhere.

achow101 commented on src/script/descriptor.cpp:
```diff
@@ -275,29 +288,37 @@ class BIP32PubkeyProvider final : public PubkeyProvider
+        KeyOriginInfo info_out;
+        CKeyID keyid = m_extkey.pubkey.GetID();
+        std::copy(keyid.begin(), keyid.begin() + sizeof(info_out.fingerprint), info_out.fingerprint);
+        info_out.path = m_path;
```
Added

achow101 committed:
Cache parent xpub inside of BIP32PubkeyProvider

Optimize Expand by having BIP32PubkeyProvider also cache the parent
(or only) xpub within itself. Since Expand does not provide a read
cache, it is useful to internally cache this xpub to avoid re-deriving
the same xpub.

achow101 commented:
I've added another commit to internally cache that parent xpub inside of `BIP32PubkeyProvider` so the `Expand` doesn't constantly re-derive the same xpub.

Sjors reviewed (commented):
~That last commit may not be necessary if `Topup()` tries `ExpandFromCache` first: https://github.com/bitcoin/bitcoin/pull/16528#discussion_r386015063 (but if I'm wrong about that, then it's fine to keep it).~  _(nvm, there are plenty of other places where Expand() is used, e.g. `importdescriptors`, `deriveaddresses`)_

Sjors commented on src/script/descriptor.cpp:
```diff
@@ -275,29 +289,57 @@ class BIP32PubkeyProvider final : public PubkeyProvider
+                if (!read_cache->GetCachedExtPubKey(parent_info, parent)) return false;
+                assert(m_derive == DeriveType::UNHARDENED);
+                assert(parent.Derive(extkey, pos));
+            }
```
```
// If we don't have a read_cache, we have a write_cache
```

Sjors reviewed (approved):
Code review ACK 384562930a48474f337224b0355b6fd08784c375 modulo assert side effects. Above and below some last suggestions to improve readability.

achow101 commented on src/script/descriptor.cpp:
```diff
@@ -275,29 +289,57 @@ class BIP32PubkeyProvider final : public PubkeyProvider
     bool IsRange() const override { return m_derive != DeriveType::NO; }
     size_t GetSize() const override { return 33; }
-    bool GetPubKey(int pos, const SigningProvider& arg, CPubKey* key, KeyOriginInfo& info) const override
+    bool GetPubKey(int pos, const SigningProvider& arg, CPubKey& key, KeyOriginInfo& info, const DescriptorCache* read_cache = nullptr, DescriptorCache* write_cache = nullptr) override
```
I don't think exclusive use is really needed or enforced.

Sjors replied:
All tests pass if I put the assert there. That implies we're not testing mixed use.

achow101 replied:
Yes, it is currently not possible to have mixed use, but I don't think we should enforce non-mixed use.

Sjors replied:
Maybe add the assert, with a comment above it along the lines of `// Mixed use of read_cache and write_cache should be possible, but is currently untested`?

achow101 replied:
I really don't think it is necessary to have an assert here.

instagibbs replied:
It's gotta be documented at least.

achow101 replied:
I've mentioned it in the comment for `GetPubKey` in `PubkeyProvider`.

achow101 commented on src/script/descriptor.cpp:
```diff
@@ -275,29 +289,57 @@ class BIP32PubkeyProvider final : public PubkeyProvider
+                // Try to get the derivation parent
+                if (!read_cache->GetCachedExtPubKey(parent_info, parent)) return false;
+                assert(m_derive == DeriveType::UNHARDENED);
+                assert(parent.Derive(extkey, pos));
```
Removed the asserts then.

achow101 replied:
I've added asserts in that shouldn't have side effects.

achow101 commented on src/script/descriptor.cpp:
```diff
@@ -291,35 +291,47 @@ class BIP32PubkeyProvider final : public PubkeyProvider
         if (read_cache) {
-            if (!read_cache->GetCachedExtPubKey(info_out, extkey)) return false;
+            if (!read_cache->GetCachedExtPubKey(info_out, extkey)) {
+                if (m_derive == DeriveType::HARDENED) return false;
```
Done

achow101 commented on src/script/descriptor.cpp:
```diff
@@ -291,35 +291,47 @@ class BIP32PubkeyProvider final : public PubkeyProvider
+                if (m_derive == DeriveType::HARDENED) return false;
+                // Try to get the derivation parent
+                if (!read_cache->GetCachedExtPubKey(parent_info, parent)) return false;
+                assert(m_derive == DeriveType::UNHARDENED);
```
Done

achow101 commented on src/script/descriptor.cpp:
```diff
@@ -275,29 +289,57 @@ class BIP32PubkeyProvider final : public PubkeyProvider
+        key = extkey.pubkey;
+        if (!m_cached_xpub.pubkey.IsValid() && m_derive != DeriveType::HARDENED) m_cached_xpub = parent;
+        if (write_cache && !read_cache) {
+            write_cache->CacheExtPubKey(info_out, extkey, true);
```
Done. Removed `skip_unhardened`.

Sjors commented:
Code review re-ACK 1373d9d885fdc671a7538f9ea6410582bcaee1d5

instagibbs commented:
The logic seems correct, but only once I did my own little refactorings to make it more understandable what everything was doing: https://github.com/instagibbs/bitcoin/commit/9c169165cbfb1565c28b8969733351e5611db6ac

I'm willing to ACK the logic as-is, but I find it extremely opaque and would immediately open up a PR with these changes.

achow101 committed:
Rename BIP32PubkeyProvider.m_extkey to m_root_extkey

Renaming clarifies that m_extkey is actually the root
extkey that keys are derived from.

achow101 commented:
I've pulled in @instagibbs suggested changes with only one small modification.

instagibbs commented:
code review ACK https://github.com/bitcoin/bitcoin/pull/18204/commits/1c20f259cf8f406254989d4ecca1c52d858209ae

thanks for pulling in most the suggestions!

sipa commented on src/script/keyorigin.h:
```diff
@@ -18,6 +18,21 @@ struct KeyOriginInfo
 
+    friend bool operator<(const KeyOriginInfo& a, const KeyOriginInfo& b)
+    {
+        for (int i = 0; i < 4; ++i) {
```
You can probably write this more readably as:

```
if (std::lexicographic_compare(std::begin(a.fingerprint), std::end(a.fingerprint), std::begin(b.fingerprint), std::end(b.fingerprint)) return true;
if (std::lexicographic_compare(std::begin(b.fingerprint), std::end(b.fingerprint), std::begin(a.fingerprint), std::end(a.fingerprint)) return false;
```

etc., which also avoids the magic 4 number.

achow101 replied:
Done

sipa commented on src/script/keyorigin.h:
```diff
@@ -18,6 +18,21 @@ struct KeyOriginInfo
+            if (a.fingerprint[i] < b.fingerprint[i]) return true;
+            if (a.fingerprint[i] > b.fingerprint[i]) return false;
+        }
+        unsigned int min = std::min(a.path.size(), b.path.size());
```
Same here:

```
if (std::lexicographic_compare(std::begin(a.path), std::end(a.path), std::begin(b.path), std::end(b.path)) return true;
if (std::lexicographic_compare(std::begin(b.path), std::end(b.path), std::begin(a.path), std::end(a.path)) return false;
```

Which also avoids manually comparing lengths.

achow101 replied:
Done

sipa commented on src/script/descriptor.cpp:
```diff
@@ -246,18 +246,19 @@ enum class DeriveType {
 {
-    CExtPubKey m_extkey;
+    // Root xpub, path, and final derivation step type being used, if any
+    CExtPubKey m_root_extkey;
```
It's not necessarily the root, right? For example in `[012345678/0']xpub.../0/1` this variable would correspond to the 2nd level intermediary node.

achow101 replied:
Sure, but it's easier to just call this the root and it's the root key that we do all our later derivation from.

achow101 replied:
I think there's enough bikeshedding in this pr :/

sipa commented on src/script/descriptor.h:
```diff
@@ -13,6 +13,17 @@
+// Cache for pre-computed descriptor things
+class DescriptorCache {
+private:
+    std::map<KeyOriginInfo, CExtPubKey> m_xpubs;
```
Shouldn't this be a multimap? There can be collisions in the fingerprint + all subsequent path elements.

achow101 replied:
I think that's extremely unlikely. It would require several thousand xpubs in the descriptor itself. I think we have a few other places where we assume origin info are unique.

sipa replied:
With 93 entries there is a 1 in a million chance for a collision... I agree it's low, but it's not so low that it is negligible, and certainly not so low that one couldn't grind a descriptor to intentionally cause a collision.

This makes me wonder if using KeyOriginInfo as key is the right approach. In theory, it should be sufficient to just key by the index of the key expression in the descriptor, right? As the descriptor is constant, and the cache is specific to one descriptor, that would uniquely identify what the entry is for.

instagibbs replied:
>index of the key expression in the descriptor

the what?

sipa replied:
Every descriptor consists of some functions that combine the output of key expressions (pubkeys, xpubs, origin info, derivation paths, ...). We can assign the successive key expressions numbers 0, 1, 2, ... and use those index positions as key in the map (which would then become a std::map<int, CExtPubKey>).

achow101 replied:
I don't really like using the position to index into the cache. It requires the `PubkeyProviders` to know their position within the descriptor when it really doesn't matter to them.

sipa replied:
I think that's not a useful argument. If the current way the code works makes it hard to implement the right solution, the code should be changed. It seems to me that using position is the more concise, exact, and efficient way of doing it, and ignoring current code architecture it also seems the simplest, so I think it would be preferable.

You're right that a multimap wouldn't work. Using a KeyOriginInfo to identify the record is inherently lossy - a better solution is just not throwing it away at all.

sipa commented on src/script/descriptor.cpp:
```diff
@@ -183,9 +187,9 @@ class OriginPubkeyProvider final : public PubkeyProvider
+    bool GetPubKey(int pos, const SigningProvider& arg, CPubKey& key, KeyOriginInfo& info, const DescriptorCache* read_cache = nullptr, DescriptorCache* write_cache = nullptr) const override
     {
-        if (!m_provider->GetPubKey(pos, arg, key, info)) return false;
+        if (!m_provider->GetPubKey(pos, arg, key, info, read_cache, write_cache)) return false;
```
Shouldn't the entry with overridden origin info be stored in the write cache here?

achow101 replied:
I don't think it would be useful to do that since we aren't reading from the cache here.

achow101 committed:
Track the index of the key expression in PubkeyProvider

achow101 commented:
I've changed the cache back to using position, but this time with `unordered_map`s instead of a vector. This requires using two maps, one for the derived keys, and one for the parent xpubs.

Also dropped the commit for `GetNotCached`.

Sjors reviewed (commented):
New approach-ACK.

Nit: "clariiesy" in 26f8ddf3183843d6e8a9ec62378010aaecf8edfe commit description

You might be able to avoid the complexity of e7e8333 `Cache parent xpub inside of BIP32PubkeyProvider`, by unifying `Expand()` and `ExpandFromCache()`. They could take single `read_write_cache` argument. You could then move the cache update code from https://github.com/bitcoin/bitcoin/pull/16528/files?file-filters%5B%5D=.cpp#diff-5462ceb8a760a507152ab8b76bd48774R1690-R1719 to `Expand()`. When no cache argument is provided, fall back to an internal cache.

instagibbs commented on src/script/descriptor.cpp:
```diff
@@ -272,37 +297,77 @@ class BIP32PubkeyProvider final : public PubkeyProvider
+        parent_info.path = m_path;
+
+        // Info of the derived key itself which is copied out upon successful completion
+        KeyOriginInfo child_info_out_tmp = parent_info;
```
apologies for re-litigating variable naming after making you change it once: it's either child or "parent" depending on the derivation, same with `child_extkey` and `child_info_out`, maybe just rename everything `final_*`?

Reviewing the latest changes made me realize my folly of naming it this way.

achow101 replied:
Done

instagibbs commented on src/script/descriptor.cpp:
```diff
@@ -789,30 +835,30 @@ std::unique_ptr<PubkeyProvider> ParsePubkey(const Span<const char>& sp, bool per
 
 /** Parse a script in a particular context. */
-std::unique_ptr<DescriptorImpl> ParseScript(Span<const char>& sp, ParseScriptContext ctx, FlatSigningProvider& out, std::string& error)
+std::unique_ptr<DescriptorImpl> ParseScript(uint32_t key_exp_index, Span<const char>& sp, ParseScriptContext ctx, FlatSigningProvider& out, std::string& error)
```
In what situations would you like to start with non-0 index?

achow101 replied:
None, but this is recursive.

instagibbs replied:
Ah! Duh.

achow101 replied:
I was intending for this to be generic so that it would work with future changes to descriptors, including miniscript. So I think I will leave it as is.

instagibbs reviewed (commented):
new approach ACK, changes make sense

achow101 commented on src/script/descriptor.h:
```diff
@@ -13,6 +13,46 @@
+// Cache for pre-computed descriptor things
+class DescriptorCache {
+private:
+    std::unordered_map<uint32_t, ExtPubKeyMap> m_derived_xpubs; // Map key expression indexes to a map of derivation indexes to xpubs
```
I think that's really inefficient. In most cases, that would mean a map for every single derived xpub, and that map just has one element. In the current mapping, it's only two maps.

instagibbs replied:
I found the code as-is completely understandable fwiw. Maybe suggest a renaming/rewording somewhere?

instagibbs replied:
maybe just make it more visual: `map: key expression index -> key derivation index -> xpub`

achow101 replied:
I've clarified the comment.

achow101 commented on src/script/descriptor.h:
```diff
@@ -13,6 +13,46 @@
 
+using ExtPubKeyMap = std::unordered_map<uint32_t, CExtPubKey>;
+
+// Cache for pre-computed descriptor things
```
Done

achow101 commented on src/script/descriptor.h:
```diff
@@ -13,6 +13,46 @@
+    /** Retrieve a cached parent xpub
+     *
+     * @param[in] key_exp_pos Position of the key expression within the descriptor
+     * @param[in] der_index Derivation index of the xpub
```
Fixed

achow101 commented on src/script/descriptor.h:
```diff
@@ -13,6 +13,46 @@
+     *
+     * @param[in] key_exp_pos Position of the key expression within the descriptor
+     * @param[in] der_index Derivation index of the xpub
+     * @param[in] xpub The CExtPubKey to cache
```
Done

achow101 commented on src/script/descriptor.h:
```diff
@@ -13,6 +13,46 @@
+     *
+     * @param[in] key_exp_pos Position of the key expression within the descriptor
+     * @param[in] der_index Derivation index of the xpub
+     * @param[in] xpub The CExtPubKey to cache
```
Done

achow101 commented on src/script/descriptor.cpp:
```diff
@@ -150,10 +150,21 @@ typedef std::vector<uint32_t> KeyPath;
 struct PubkeyProvider
 {
+protected:
+    //! Index of this key expression in the descriptor
```
I added an example to the comment.

achow101 commented on src/script/descriptor.cpp:
```diff
@@ -150,10 +150,21 @@ typedef std::vector<uint32_t> KeyPath;
 {
+protected:
+    //! Index of this key expression in the descriptor
+    uint32_t m_exp_index;
```
Done

instagibbs commented on src/script/descriptor.cpp:
```diff
@@ -272,37 +298,77 @@ class BIP32PubkeyProvider final : public PubkeyProvider
+        if (m_derive == DeriveType::HARDENED) final_info_out_tmp.path.push_back(((uint32_t)pos) | 0x80000000L);
+
+        // Derive keys or fetch them from cache
+        CExtPubKey child_extkey = m_root_extkey;
```
Was also thinking changing this to `final_extkey` as well for same reason.

achow101 replied:
Done

instagibbs reviewed (approved):
code review ACK https://github.com/bitcoin/bitcoin/pull/18204/commits/196c9765eaa88120964cfb1a81362f37a24ced9b

instagibbs commented:
code review re-ACK https://github.com/bitcoin/bitcoin/pull/18204/commits/d7b241109f74b3c1c407929930e7d035cdcf60c9

Sjors commented on src/script/descriptor.cpp:
```diff
@@ -345,10 +345,11 @@ class BIP32PubkeyProvider final : public PubkeyProvider
             // Only cache parent if there is any unhardened derivation
             if (m_derive != DeriveType::HARDENED) {
                 write_cache->CacheParentExtPubKey(m_expr_index, parent_extkey);
+            } else if (child_info_out.path.size() > 0) {
```
e4262c45655b4f79ddfd9905db66dc44b8f5148c doesn't compile (`child_info_out`  is  renamed in In d7b241109f74b3c1c407929930e7d035cdcf60c9 `to `child_info_out`)

achow101 replied:
Fixed

Sjors reviewed (commented):
Code review ACK d7b241109f74b3c1c407929930e7d035cdcf60c9 modulo intermediate commit compile issue.

achow101 commented on src/script/descriptor.h:
```diff
@@ -13,6 +13,45 @@
 
+using ExtPubKeyMap = std::unordered_map<uint32_t, CExtPubKey>;
+
+// Cache for single descriptor's derived extended pubkeys
```
Done

achow101 commented on src/script/descriptor.h:
```diff
@@ -13,6 +13,45 @@
+// Cache for single descriptor's derived extended pubkeys
+class DescriptorCache {
+private:
+    std::unordered_map<uint32_t, ExtPubKeyMap> m_derived_xpubs; // Map key expression index -> map of (key derivation index -> xpub)
```
Done

Sjors replied:
It needs to be on the line above, otherwise this happens:
<img width="808" alt="Schermafbeelding 2020-03-07 om 15 53 36" src="https://user-images.githubusercontent.com/10217/76145663-d47e7780-608b-11ea-8f09-8d4399a0d315.png">

achow101 replied:
Fixed

achow101 commented on src/script/descriptor.h:
```diff
@@ -13,6 +13,45 @@
+class DescriptorCache {
+private:
+    std::unordered_map<uint32_t, ExtPubKeyMap> m_derived_xpubs; // Map key expression index -> map of (key derivation index -> xpub)
+    ExtPubKeyMap m_parent_xpubs; // Map key expression index -> parent xpub
```
Done

achow101 commented on src/script/descriptor.h:
```diff
@@ -13,6 +13,45 @@
+     */
+    bool GetCachedDerivedExtPubKey(uint32_t key_exp_pos, uint32_t der_index, CExtPubKey& xpub) const;
+
+    const ExtPubKeyMap GetCachedParentExtPubKeys() const;
```
Done

achow101 commented on src/script/descriptor.h:
```diff
@@ -13,6 +13,45 @@
+    bool GetCachedDerivedExtPubKey(uint32_t key_exp_pos, uint32_t der_index, CExtPubKey& xpub) const;
+
+    const ExtPubKeyMap GetCachedParentExtPubKeys() const;
+    const std::unordered_map<uint32_t, ExtPubKeyMap> GetCachedDerivedExtPubKeys() const;
```
Done

Sjors commented:
re-ACK 31304fc4f8f0d7870794d8c675831d8c73c959cb

instagibbs commented:
code review re-re-ACK https://github.com/bitcoin/bitcoin/pull/18204/commits/09e25071f40c564af08a1386c39c4f2d8eb484b6

Sjors reviewed (approved):
re-ACK 09e25071f40c564af08a1386c39c4f2d8eb484b6
